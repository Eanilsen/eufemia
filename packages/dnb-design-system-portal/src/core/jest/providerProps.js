/**
 * Test Prop/Store Provider
 *
 */

import React from 'react'
import PropTypes from 'prop-types'
import { Router } from 'react-static'
import { Provider } from 'mobx-react'

import { MobxIntlProvider, localeStore } from '../startup/locales'
import history from '../routing/History'
// import stores from '../../dm/stores'

import extend from 'lodash.merge'

// make Object.merge global aviable
Object.merge = (...args) => {
  args.unshift({})
  return extend.apply(extend, args)
} // I used merge before, but the "array-includes-with-glob" is not ES5 code in the npm package, so this causes troubles on the build / UglifyJsPlugin process
Object.extend = extend

// make sure the lang is set to en

const ProviderSimulation = props => {
  localeStore.value = props.lang || 'en'
  return (
    <Provider
      locale={localeStore}
      // {...stores}
    >
      <MobxIntlProvider>
        <Router history={history}>{props.children}</Router>
      </MobxIntlProvider>
    </Provider>
  )
}
ProviderSimulation.propTypes = {
  lang: PropTypes.string,
  children: PropTypes.node.isRequired
}
ProviderSimulation.defaultProps = {
  lang: 'en'
}

export default {
  Provider: ProviderSimulation,
  stores: Object.assign(
    { history: history, locale: localeStore }
    // , stores
  )
}
