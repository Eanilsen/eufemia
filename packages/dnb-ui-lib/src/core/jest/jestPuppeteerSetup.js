/**
 * Jest Setup for Screenshot testing
 *
 */

const chalk = require('chalk')
const puppeteer = require('puppeteer')
const mkdirp = require('mkdirp')
const path = require('path')
const fs = require('fs')
const isCI = require('is-ci')
const liveServer = require('live-server')
const detectPort = require('detect-port')
const {
  DIR,
  testScreenshotOnHost,
  testScreenshotOnPort
} = require('./jestSetupScreenshots')

module.exports = async function() {
  console.log(chalk.green('Setup Puppeteer'))
  const portIsAvailable = await detectPort(testScreenshotOnPort)
  if (testScreenshotOnPort === portIsAvailable) {
    const params = {
      port: testScreenshotOnPort,
      host: testScreenshotOnHost,
      root: '../dnb-design-system-portal/public',
      open: false,
      watch: '',
      quiet: isCI,
      wait: 10e3
    }
    liveServer.start(params)
    // liveServer.server.addListener('listening', () => () => {
    // })
  }

  const browser = await puppeteer.launch()
  // store the browser instance so we can teardown it later
  // this global is only available in the teardown but not in TestEnvironments
  global.__BROWSER_GLOBAL__ = browser

  // use the file system to expose the wsEndpoint for TestEnvironments
  mkdirp.sync(DIR)
  fs.writeFileSync(path.join(DIR, 'wsEndpoint'), browser.wsEndpoint())
}
