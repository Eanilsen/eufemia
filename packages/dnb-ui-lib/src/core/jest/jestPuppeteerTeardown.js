/**
 * Jest Setup for Screenshot testing
 *
 */

const fs = require('fs-extra')
const packpath = require('packpath')
const path = require('path')
const tar = require('tar')
const chalk = require('chalk')
const rimraf = require('rimraf')
const isCi = require('is-ci')
const liveServer = require('live-server')
import {
  commitToBranch,
  getCurrentBranchName
} from '../../../scripts/prepub/commitToBranch'
const { DIR } = require('./jestSetupScreenshots').config

module.exports = async function() {
  if (isCi) {
    console.log(chalk.yellow('Teardown Puppeteer.'))
  }

  if (liveServer.shutdown) {
    liveServer.shutdown()
  }
  // close the browser instance
  await global.__BROWSER_GLOBAL__.close()

  // commit a tar of the reports if we are on a CI
  if (isCi) {
    const reportPath = path.resolve(
      packpath.self(),
      './jest-screenshot-report'
    )
    if (fs.existsSync(reportPath)) {
      const branchName = await getCurrentBranchName()
      await tar.create(
        {
          gzip: true,
          file: `./reports/${branchName}-jest-screenshot-report.tgz`
        },
        ['./jest-screenshot-report']
      )
      await commitToBranch({
        skipCI: true,
        requiredBranch: ['develop', 'rc/*'],
        what: 'reports',
        filePathsWhitelist: ['jest-screenshot-report.tgz'],
        isNotAFeature: ['jest-screenshot-report.tgz']
      })
    }
  }

  // clean-up the wsEndpoint file
  rimraf.sync(DIR)
}
