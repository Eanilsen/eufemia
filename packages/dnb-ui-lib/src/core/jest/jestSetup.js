/**
 * Default Component Test Setup
 *
 */

import '../startup/required'

import { axe, toHaveNoViolations } from 'jest-axe'
// TODO: fakeDataForProps gets not exported properly
// import fakeProps, { fakeDataForProps } from 'react-fake-props'
// This is the reason, why we use our own "copy"
import fakeProps, {
  fakeDataForProps as _fakeDataForProps
} from './react-fake-props'
import { mount, render, shallow } from './enzyme'

import * as reactDocs from 'react-docgen'
import ReactDOMServer from 'react-dom/server'
import fs from 'fs-extra'
import onceImporter from 'node-sass-once-importer'
import path from 'path'
import puppeteer from 'puppeteer'
import sass from 'node-sass'
import { setupJestScreenshot } from 'jest-screenshot'
import { toBeType } from 'jest-tobetype'
import toJson from 'enzyme-to-json'

export {
  fakeProps, // we have also our own replacement function called "fakeAllProps"
  // fakeDataForProps,
  shallow,
  mount,
  render,
  toJson,
  axe,
  toHaveNoViolations,
  setupJestScreenshot
}

expect.extend({ toBeType })
expect.extend(toHaveNoViolations)

const fakeDataForProps = (props, options) => {
  // there is a bug in "react-docgen"
  // to make sure we don't return enum strings with an \'...\' inside, we remove it here
  for (let i in props) {
    if (props[i].type.name === 'enum' && props[i].type.value) {
      if (Array.isArray(props[i].type.value))
        props[i].type.value = props[i].type.value.map(({ value }) => ({
          // no, we dont want this in a string
          value: value.replace(new RegExp("'", 'g'), '')
        }))
    }
  }
  return _fakeDataForProps(props, options)
}

// Note: replace this code later, once "react-fake-props" is exporting fakeDataForProps properly
export const fakeAllProps = (file, options) => {
  const source = fs.readFileSync(file, 'utf-8')
  const componentInfo = reactDocs.parse(
    source,
    reactDocs.resolver.findAllComponentDefinitions
  )
  return componentInfo.props
    ? fakeDataForProps(componentInfo.props, options)
    : // in case we use findAllComponentDefinitions
    // we have to walk thouh all the results
    Array.isArray(componentInfo)
    ? componentInfo.reduce(
        (acc, cur) => ({
          ...acc,
          ...fakeDataForProps(cur.props, options)
        }),
        {}
      )
    : {}
}

export const loadScss = file => {
  try {
    const sassResult = sass.renderSync({
      file: file,
      includePaths: [path.resolve(__dirname, '../../style/core/')],
      importer: [onceImporter()]
    })
    return String(sassResult.css)
  } catch (e) {
    console.log('Error', e)
  }
}

export const setupPageScreenshot = (options = { timeout: 10e3 }) => {
  // just setup this one time
  if (global.browser) return

  setupJestScreenshot(
    options
    // {
    // // ...{
    // //   // detectAntialiasing: true, // Whether to attempt to detect antialiasing and ignore related changes when comparing both images.
    // //   // pixelThresholdRelative: 0, // If specified, jest-screenshot will fail if more than the specified relative amount of pixels are different from the snapshot. When setting this to 0.5 for example, more than 50% of the pixels need to be different for the test to fail.
    // //   // colorThreshold: 1 // A number in the range from 0 to 1 describing how sensitive the comparison of two pixels should be.
    // //   // colorThreshold: 0
    // // },
    // // ...options
    // }
  )

  beforeAll(async done => {
    try {
      const browser = await puppeteer.launch({
        args: ['--no-sandbox', '--disable-setuid-sandbox']
      })

      global.browser = browser
    } catch (err) {
      console.error('Unable to start puppeteer.', err)
    }
    done()
  })

  afterAll(async done => {
    if (!global.browser) {
      done()
    }
    await global.browser.close()
    done()
  })

  beforeEach(async done => {
    if (!global.browser || !global.browser.newPage) {
      done()
    }
    global.page = await global.browser.newPage()

    // await global.page.setViewport({ width: 1920, height: 1080 })

    // some optimisations?
    // await global.page.setRequestInterception(true)
    // global.page.on('request', req => {
    //   switch (req.resourceType()) {
    //     case 'image':
    //       // case 'stylesheet':
    //       // case 'script':
    //       // case 'font':
    //       // case 'xhr':
    //       // case 'eventsource':
    //       // case 'document':
    //       req.abort()
    //       break
    //     default:
    //       req.continue()
    //   }
    // })

    done()
  })

  // make sure jest is waiting for 10 sec
  jest.setTimeout(options.timeout)
}

export const loadImage = async imagePath => {
  return await fs.readFile(path.resolve(imagePath))
}

export const axeComponent = async Component => {
  const html = ReactDOMServer.renderToStaticMarkup(Component)
  return await axe(html)
}

// global.shallow = enzyme.shallow
// global.render = enzyme.render
// global.mount = enzyme.mount

function jestSetup() {}

export default jestSetup
