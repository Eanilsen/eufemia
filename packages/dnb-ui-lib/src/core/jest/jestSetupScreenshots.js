/**
 * Jest Setup for Screenshot testing
 * github.com/GoogleChrome/puppeteer/blob/master/docs/api.md
 */

const path = require('path')
const os = require('os')
const isCI = require('is-ci')
const { setupJestScreenshot } = require('jest-screenshot')

const testScreenshotOnHost = '127.0.0.1'
// use same port as the local dev setup, this way we can test from the dev setup as well
const testScreenshotOnPort = 8000
module.exports.testScreenshotOnHost = testScreenshotOnHost
module.exports.testScreenshotOnPort = testScreenshotOnPort
module.exports.DIR = path.join(os.tmpdir(), 'jest_puppeteer_global_setup')

const openPageScreenshot = (url, { cachePages } = {}) =>
  new Promise(async (resolve, reject) => {
    try {
      const gotoUrl = `http://${testScreenshotOnHost}:${testScreenshotOnPort}/${url}`.replace(
        /\/\//g,
        '/'
      )

      if (global.gotoUrl !== gotoUrl) {
        global.page = await global.__BROWSER__.newPage()
        global.page.setViewport({
          width: 800,
          height: 600,
          isMobile: false,
          hasTouch: false,
          isLandscape: false,
          deviceScaleFactor: 1
        })
        await global.page.goto(gotoUrl)
        if (cachePages) {
          global.gotoUrl = gotoUrl
        }
      }

      resolve(global.page)
    } catch (e) {
      reject(e)
    }
  })
module.exports.openPageScreenshot = openPageScreenshot

module.exports.testPageScreenshot = ({
  url = null,
  cachePages = !isCI,
  page = null,
  selector,
  simulate = null,
  transformElement = null
} = {}) =>
  new Promise(async (resolve, reject) => {
    try {
      if (url && !page) {
        page = await openPageScreenshot(url, { cachePages })
        await page.waitForSelector(selector)
      }

      const element = await page.$(selector)

      if (transformElement) {
        await transformElement(element)
      }

      let delay = 0

      if (simulate) {
        switch (simulate) {
          case 'hover':
            {
              await element.hover()
            }
            break

          case 'active':
            {
              // make a delayed click, no await. Else we get only a release state
              delay = 500
              element.click({
                delay // move the mouse
              })
            }
            break

          case 'focus':
            {
              await element.press('Tab') // to simulate pressing tab key before focus
              await element.focus()
            }
            break
        }
      }

      const screenshot = await element.screenshot()

      // just to make sure we dont resolve, before the delayed click happened
      if (delay > 0) {
        await wait(delay)
      }

      if (!cachePages) {
        page.close()
      }

      resolve(screenshot)
    } catch (e) {
      reject(e)
    }
  })

module.exports.setupPageScreenshot = ({ timeout, ...rest } = {}) => {
  if (timeout) {
    jest.setTimeout(timeout)
  }
  setupJestScreenshot(rest)
}

const wait = t => new Promise(r => setTimeout(r, t))
