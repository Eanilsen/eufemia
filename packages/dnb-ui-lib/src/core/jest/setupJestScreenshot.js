/**
 * Jest Setup for Screenshot testing
 *
 */

const isCI = require('is-ci')
const { setupJestScreenshot } = require('jest-screenshot')

setupJestScreenshot({
  pixelThresholdRelative: isCI ? 0.15 : 0.015
})
