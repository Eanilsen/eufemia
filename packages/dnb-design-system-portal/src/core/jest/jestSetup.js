/**
 * Default Component Test Setup
 *
 */

import fakeProps from 'react-fake-props'
import * as enzyme from './enzyme'
import enzymeToJson from 'enzyme-to-json'
import * as emotion from 'emotion'
import { createSerializer } from 'jest-emotion'
import providerProps from './providerProps'
import '../startup/required'

expect.addSnapshotSerializer(
  createSerializer(emotion, {
    classNameReplacer(className, index) {
      return `jest-class-name-${index}`
    }
  })
)

module.exports = {
  ...enzyme,
  fakeProps,
  enzymeToJson,
  providerProps,
  toJson: enzymeToJson
}
