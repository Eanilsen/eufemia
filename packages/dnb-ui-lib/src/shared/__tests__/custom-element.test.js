/**
 * Custom element tests
 *
 */

import React from 'react'

import { registerElement, registeredElements } from '../custom-element'
import { toJson } from '../../core/jest/jestSetup'

// IMPORTANT: The tests requires that the custom element is polyfilled by using "document-register-element"

const bodyElements = window.document.getElementsByTagName('body')
const bodyElement = bodyElements[0]
const CustomElementComponentContent = 'I have a rendered content'
bodyElement.insertAdjacentHTML(
  'afterbegin',
  `<custom-element observed-attribute="${CustomElementComponentContent}">Startup Content</custom-element>`
)

class CustomElementComponent extends React.Component {
  state = {}
  static getDerivedStateFromProps(props, state) {
    state.content = props['observed-attribute']
    return state
  }
  constructor(props) {
    super(props)
    this.state.content = CustomElementComponentContent
  }
  render() {
    return <div>{this.state.content}</div>
  }
}

describe('"registerElement" should', () => {
  registerElement('custom-element', CustomElementComponent, [
    'observed-attribute'
  ])
  const customElement = window.document.getElementsByTagName(
    'custom-element'
  )

  it('have "registeredElements" with more than one element', () => {
    expect(registeredElements.length).toBeGreaterThanOrEqual(1)
  })
  it('have "registeredElements" with the correct tagName registered', () => {
    expect(registeredElements[0]).toBe('custom-element')
  })

  it('have a "custom-element" tag in the jsdom', () => {
    expect(customElement.length).toBe(1)
  })

  it('have a "custom-element" with the correct element value', () => {
    expect(customElement[0].textContent).toBe(
      CustomElementComponentContent
    )
  })

  it('have a "custom-element" with the correct attribute value', () => {
    expect(customElement[0].getAttribute('observed-attribute')).toBe(
      CustomElementComponentContent
    )
  })

  it('have a "custom-element" with the correct element content value, even after changing the attribute value', () => {
    const newAttributeValue = 'new-value'
    customElement[0].setAttribute('observed-attribute', newAttributeValue)
    expect(customElement[0].textContent).toBe(newAttributeValue)
  })

  it('handle "custom event" as expected', () => {
    const eventCallback = jest.fn()
    const eventParams = { value: 123 }
    const eventId = customElement[0].addEvent('on_change', eventCallback)

    // call the event one time
    customElement[0].fireEvent('on_change', eventParams)
    customElement[0].removeEvent(eventId)

    // should not trigger the event
    customElement[0].fireEvent('on_change', eventParams)

    // expect the event to be called once
    expect(eventCallback.mock.calls.length).toBe(1)

    // and check for some more criterias
    expect(eventCallback.mock.calls[0][0]).toBe(eventParams)
    const eventId2 = customElement[0].addEvent('on_change', jest.fn())
    expect(eventId).not.toBe(eventId2)
  })

  it('handle "native event" (CustomEvent) as expected', () => {
    const eventCallback = jest.fn()
    const detail = { value: 123 }
    const customEvent = new CustomEvent('on_change', {
      bubbles: true,
      cancelable: true,
      detail
    })
    customElement[0].addEventListener('on_change', eventCallback)

    // call the event one time
    customElement[0].dispatchEvent(customEvent)
    customElement[0].removeEventListener('on_change', eventCallback)

    // should not trigger the event
    customElement[0].dispatchEvent(customEvent)

    // expect the event to be called once
    expect(eventCallback.mock.calls.length).toBe(1)
    expect(eventCallback.mock.calls[0][0].detail).toBe(detail)
  })

  it('have to match snapshot', () => {
    expect(toJson(customElement)).toMatchSnapshot()
  })
})
