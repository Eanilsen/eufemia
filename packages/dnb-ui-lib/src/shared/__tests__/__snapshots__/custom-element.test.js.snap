// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"registerElement" should have to match snapshot 1`] = `undefined`;
